lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -752,12 +770,68 @@ selectLatestCheckpoint wid = fmap entityVal <$>
 selectUTxO
     :: Checkpoint
     -> SqlPersistT IO [UTxO]
-selectUTxO (Checkpoint wid sl _parent _bh) = fmap entityVal <$>
+selectUTxO (Checkpoint wid _sl _parent _bh) = fmap entityVal <$>
     selectList
         [ UtxoWalletId ==. wid
-        , UtxoSlot ==. sl
+        , UtxoSlotSpent ==. Nothing
         ] []
 
+insertWhenNew
+    :: W.WalletId
+    -> UTxO
+    -> SqlPersistT IO ()
+insertWhenNew wid utxo@(UTxO _ _ _ inpId inpIx outAddr outAmt) = do
+    utxos <- selectList [ UtxoWalletId ==. wid
+                        , UtxoInputId ==. inpId
+                        , UtxoInputIndex ==. inpIx
+                        , UtxoOutputAddress ==. outAddr
+                        , UtxoOutputCoin ==. outAmt
+                        ] []
+    when (null utxos) $ insert_ utxo
+
+addSpentUtxos
+    :: [TxMeta]
+    -> [TxIn]
+    -> SqlPersistT IO ()
+addSpentUtxos metas txins = do
+    let outgoingAndInLedger =
+            filter (\(TxMeta _ _ st dir _ _) ->
+                        st == W.InLedger && dir == W.Outgoing) metas
+    mapM_ (updateUtxosPerMeta txins) outgoingAndInLedger
+
+updateUtxosPerMeta
+    :: [TxIn]
+    -> TxMeta
+    -> SqlPersistT IO ()
+updateUtxosPerMeta txins (TxMeta txid wid _ _ sl _) = do
+    let validTxIns = filter (\(TxIn txid' _ _ _ _ ) -> txid'==txid) txins
+    mapM_ (updateUtxoSpent wid sl) validTxIns
+
+updateUtxoSpent
+    :: W.WalletId
+    -> W.SlotId
+    -> TxIn
+    -> SqlPersistT IO ()
+updateUtxoSpent wid sl (TxIn _ _ inpId inpIx _) = do
+    utxos <- selectList [ UtxoWalletId ==. wid
+                        , UtxoSlotSpent ==. Nothing
+                        , UtxoInputId ==. inpId
+                        , UtxoInputIndex ==. inpIx
+                        ] [LimitTo 1]
+    mapM_ _updateUtxoSpent utxos
```

Comment:
This should be using `updateWhere`. My model is to try to think of what SQL query would achieve the result, then translate it to persistent.

Reply:
corrected. used `updateWhere`